# pipelines.chromake.scripts.paths.get_all_fastq_related_paths { #genomake.pipelines.chromake.scripts.paths.get_all_fastq_related_paths }

```python
pipelines.chromake.scripts.paths.get_all_fastq_related_paths(cfg, mode)
```

Get the path to the files generated by the pipeline for the samples of all sequencing and projects.

## Parameters {.doc-section .doc-section-parameters}

| Name   | Type   | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | Default    |
|--------|--------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------|
| cfg    | dict   | Dict representing the configuration of an analysis with the chromake pipeline.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | _required_ |
| mode   | str    | String representing the files to get. Accepted values are:     - sequencing (raw fastq)     - multiqc_raw (multiqc report regrouping the fastqc of the raw fastq files)     - cutadapt (fastq after adapter trimming)     - fastqc_trimmed (fastqc reports after the adapter trimming)     - multiqc_trimmed (multiqc report after adapter trimming)     - bam (bam files after alignment with bowtie2)     - bam_filtered (bam files after filtering the non-standard chromosomes)     - bedgraph (bedGrapg files generated by HOMER for UCSC visualisation) | _required_ |

## Returns {.doc-section .doc-section-returns}

| Name   | Type   | Description           |
|--------|--------|-----------------------|
|        | list   | A list of file paths. |